---
title: "Intro to Transcriptomics"
author: Nicole Black, Wade Boohar, Kayla Xu
date: 07/17/22
updated: 09/26/23
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partnerâ€™s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In Transcriptomics_Tutorial.Rmd, you learned how to manipulate the SummarizedExperiment data structure, turn it into more readable dataframes, complete basic analysis using those dataframes, and performed a differential expression analysis based on age. Before saving the rna_counts, rna_clinical, and rna_genes however, we masked out any sample that was not a tumor. In this assignment, you will need to reinstantiate the rna_se object and then perform differential expression analysis based on tumor status.

*Pre-Assignment*
Use knitr function to set your working directory to your analysis_data folder in qbio_490_name.
```{r}
setwd('/home1/awbai/analysis_data')
```

Load all necessary packages and instantiate the rna_se object
```{r}
#Loaded in summarizedexperiment and tcgabiolinks
install.packages('BiocManager')
BiocManager::install('TCGAbiolinks')
library(TCGAbiolinks)
install.packages('SummarizedExperiment')
library(SummarizedExperiment)
BiocManager::install('DESeq2')

#Opened rna_se through a downloaded file because lack of memory 

```


*1*
Perform all data cleaning and label manipulation steps (hint: applying colnames and rownames to make the dataframes more readable) except for masking out the patient with an NA for age and the "Solid Tissue Normal" sample types.
```{r}
rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)

#Change rna_genes from s4 object to data frame
rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)

#Creating a new dataframe with data from unstranded 
rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)

#Making row names informative for rna_genes
rownames(rna_genes) <- rna_genes$gene_id

#Making row and column names informative for rna_counts
rownames(rna_counts) <- rna_genes$gene_id
colnames(rna_counts) <- rownames(rna_clinical)

#Subset any samples outside normal tissue from rna_clinical and rna_counts
unique(rna_clinical$definition)

normal_mask <- ifelse(rna_clinical$definition == "Metastatic", F, T)

rna_clinical <- rna_clinical[normal_mask,]

rna_counts <- rna_counts[, normal_mask]


#Removing NAs from all variables in rna_clinical
#gender_na_mask <- ifelse(is.na(rna_clinical$gender), FALSE, TRUE)
#rna_clinical <- rna_clinical[gender_na_mask, ]

vital_status_na_mask <- ifelse(is.na(rna_clinical$vital_status), FALSE, TRUE)
rna_clinical <- rna_clinical[vital_status_na_mask, ]

definition_na_mask <- ifelse(is.na(rna_clinical$definition), FALSE, TRUE)
rna_clinical <- rna_clinical[definition_na_mask,]

#Removing NAs from all variables in rna_counts

#rna_counts <- rna_counts[,gender_na_mask]

rna_counts <- rna_counts[,vital_status_na_mask]

rna_counts <- rna_counts[,definition_na_mask]

```

*2*
<<<<<<< HEAD
Determine and define your experimental design for the differential expression analysis. Control for covariance of age, gender, OR vital_status. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?)
=======
Determine and define your experimental design for the differential expression analysis. Control for covariance of at least 2 other variables. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what kind of variable do they need to be? is the data type correct? do you need to handle unknown values?)


```{r}
# Make a comparison of 1 clinical variable across all genes Ex: young vs old in age
# Categorical variable: definition
# Covariance variables: vital_status

#Removing genes with less than 10 patient samples

# use rowSums() to create a list with the total number of counts of each gene
row_sums <- rowSums(rna_counts, na.rm=TRUE)

# create a boolean mask where genes with < 10 total counts are FALSE, and genes with >= 10 total counts are TRUE  
low_counts_mask <- ifelse(row_sums >= 10, TRUE, FALSE)

# rewrite the rna_counts df, subsetting for only genes with >= 10 total counts
rna_counts <- rna_counts[low_counts_mask, ]


#update rna_genes with the low_counts_mask
rna_genes <- rna_genes$gene_id[low_counts_mask]


#Factor categorical variable and covariance variables 
rna_clinical$vital_status <- factor(rna_clinical$vital_status, levels=c('Alive', 'Dead'))
head(rna_clinical$vital_status)

#rna_clinical$gender <- factor(rna_clinical$gender, levels=c('male', 'female'))
#head(rna_clinical$gender)

rna_clinical$definition <- factor(rna_clinical$definition, levels=c('Primary solid Tumor', 'Solid Tissue Normal'))
head(rna_clinical$definition)

```

```{r}
ncol(rna_counts)
ncol(rna_clinical)

nrow(rna_counts)
nrow(rna_clinical)
```



*3*
Perform the differential expression analysis, create a volcano plot, and interpret what genes from each of the 6 sections means (assuming thresholds of -log10padj = 2 and log2FoldChange = 1)
```{r}
?DESeqDataSetFromMatrix
dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                             colData = rna_clinical,
                             design = ~vital_status + definition)

?DESeq
dds_obj <- DESeq(dds) # note: this will likely take a long time (ie 45 minutes to 2 hours)

?resultsNames
resultsNames(dds_obj)  # see what comparisons got run

# get the young vs. old comparison
#?results
results <- results(dds_obj, format = "DataFrame", contrast = c("definition", 'Primary solid Tumor', 'Solid Tissue Normal')) # this is case sensitive so be careful to match it with your age_category factors closely!



library(EnhancedVolcano)

#Volcano plot
  EnhancedVolcano(results,
    lab = rownames(results),
    x = 'log2FoldChange',
    y = 'pvalue')

  #put in somewhere different first!!!!! to not replace

```

top-right genes: These show positive change in the amount of genetic mutations between patients with a primary solid tumor or normal tissue. Being on the top indicates that these genes are statistically significant.
bottom-right genes: Statistically insignificant genes. The horizontal threshold is indicative of whether genes have statistically significant differential expression.
top-left genes: These genes are less mutated between our two populations, but are statistically significant.
bottom-left genes: Statistically insignificant genes that exhibit less genetic mutation. 
top-middle genes: There are no significant changes in gene mutation between patients with primary solid tissue or normal tissue within this region. These are statistically significant.
bottom-middle genes: Statistically insignificant genes with no changes in mutation between our two patient populations.

Save the picture of the volcano plot (using either ggsave() or right clicking and manually downloading the image and push this .Rmd and the image to GitHub)
