---
title: "R Review Project"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

```{r}
load("C:/Users/15713/Desktop/QBio490/qbio_490_angelabai/rna_se_skcm")

ncol(rna_se@colData)
nrow(rna_se@colData)
```


```{r}
#Setting working directory

knitr::opts_knit$set(root.dir = normalizePath("C:/Users/15713/Desktop/QBio490/qbio_490_angelabai/analysis_data")) 

#Downloading packages
if (!require("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install(version = "3.17")
if (!require("TCGAbiolinks", quietly = TRUE))
BiocManager::install("TCGAbiolinks")
if (!require("maftools", quietly = TRUE))
BiocManager::install("maftools")
library(BiocManager)
library(TCGAbiolinks)
library(maftools)

#Read rna_se into the enviornment 

rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)

```

```{r}
#1 Kaplan-Meyer Plotting 

#Loading packages
if (!require(survival)) {
  install.packages("survival")
}
library(survival)

if (!require(survminer)) {
  install.packages("survminer")
}
library(survminer)

if (!require(ggplot2)) {
  install.packages("ggplot2")
}
library(ggplot2)


#Create a column for metastasis 
rna_clinical$metastasis <- rna_se@colData$definition

#Clean NA from rna_metastasis
metastasis_na_mask <- !is.na(rna_clinical$metastasis)
rna_clinical <- rna_clinical[metastasis_na_mask,]

#Mask out patients who are not "Metastatic" or "Primary solid Tumor"
metastatic_mask <- ifelse(rna_clinical$metastasis == 'Metastatic', TRUE, ifelse(rna_clinical$metastasis == 'Primary solid Tumor', TRUE, FALSE))
#primary_solid_tumor_mask <- ifelse(rna_clinical$metastasis == 'Primary solid Tumor', TRUE, FALSE)

rna_clinical <- rna_clinical[metastatic_mask,]
#rna_clinical <- rna_clinical[primary_solid_tumor_mask,]


#Creating survival_time which combines information from death_days_to (patients who passed away) and last_contact_days_to (patients who are still alive)
rna_clinical$survival_time <- ifelse(rna_clinical$days_to_death == "[Not Applicable]", rna_clinical$days_to_last_follow_up, rna_clinical$days_to_death)

#Converting survival_time to numeric values 
rna_clinical$survival_time <- as.numeric(rna_clinical$survival_time)

#Creating a death_event column that contains "True" if a patient has died and "False" if a patient is still alive
rna_clinical$death_event <- ifelse(rna_clinical$vital_status == "Alive", FALSE, TRUE)

#Creating Survminer objects
survival_object <- Surv(time = rna_clinical$survival_time,
                        event = rna_clinical$death_event)

fit_object <- survfit(survival_object ~ metastasis, data = rna_clinical)

#Creating KM Plot
survplot <- ggsurvplot(fit_object,
                       pval=TRUE,
                       ggtheme = theme(plot.margin = unit(c(1,1,1,1), "cm")),
                       legend = 'right')

KM_plot <- survplot$plot + theme_bw() + theme(axis.title = element_text(size=20), 
                                              axis.text = element_text(size=16),
                                              legend.title = element_text(size=14),
                                              legend.text = element_text(size=12))

KM_plot


#Issue: The strata shows "all" instead of metastastic vs non-metastatic patients 



```





```{r}
#2 CoOncoplot

#Loading necessary packages


library(maftools)

  knitr::opts_knit$set(root.dir = normalizePath("/Users/15713/Desktop/QBio490/qbio_490_angelabai/GDCdata")) 

maf_query <- GDCquery(
    project = "TCGA-SKCM", 
    data.category = "Simple Nucleotide Variation", 
    access = "open", # we only have access to somatic mutations which are open access
    data.type = "Masked Somatic Mutation", 
    workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
)


#Create tumor_sample_barcode
rna_clinical$Tumor_Sample_Barcode <- rna_clinical$patient

#Instantiating MAF_object

#GDCdownload(maf_query)

maf <- GDCprepare(maf_query) # as long as it runs, ignore any errors

maf_object <- read.maf(maf = maf, 
                       clinicalData = rna_clinical,
                       isTCGA = TRUE)


#Separate population into two groups

metastatic_mask <- ifelse(rna_clinical$metastasis == 'Metastatic', TRUE, FALSE)
metastatic_barcode <- maf_object@clinical.data$Tumor_Sample_Barcode[metastatic_mask]

metastatic_maf <- subsetMaf(maf = maf_object,
                            tsb = metastatic_barcode)

primary_solid_tumor_barcode <- maf_object@clinical.data$Tumor_Sample_Barcode[!metastatic_mask]

primary_solid_tumor_maf <- subsetMaf(maf = maf_object,
                            tsb = primary_solid_tumor_barcode)

#Oncoplot to identify top 10 genes 
oncoplot(maf = maf_object,
         top = 10,
         borderCol = NA) 

#CoOncoplot
coOncoplot(m1 = metastatic_maf, 
           m2 = primary_solid_tumor_maf, 
           m1Name = 'Metastatic Patients', 
           m2Name = 'Nonmetastatic Patients',
           genes = c("TTN","MUC16","BRAF","DNAH5","PCLO","LRP1B","ADGRV1","ANK3","CSMD1","DNAH7"),
           borderCol = NA)


```






```{r}
#Colollipop

#Commonly mutated gene of interest: TTN

#Separate population into two groups

metastatic_mask <- ifelse(rna_clinical$metastasis == 'Metastatic', TRUE, FALSE)
metastatic_barcode <- maf_object@clinical.data$Tumor_Sample_Barcode[metastatic_mask]

metastatic_maf <- subsetMaf(maf = maf_object,
                            tsb = metastatic_barcode)

primary_solid_tumor_barcode <- maf_object@clinical.data$Tumor_Sample_Barcode[!metastatic_mask]

primary_solid_tumor_maf <- subsetMaf(maf = maf_object,
                            tsb = primary_solid_tumor_barcode)

lollipopPlot2(m1 = metastatic_maf, 
              m2 = primary_solid_tumor_maf, 
              m1_name = "Metastatic Patients",
              m2_name = "Nonmetastatic Patients",
              gene = "TTN") 




```



```{r}
#4 Contingency table for metastatic patients (DO I CHOOSE ONE GENE OR MULTIPLE)

rna_clinical$metastasis <- factor(rna_clinical$metastasis, levels=c('Metastatic', 'Primary solid Tumor'))

#Subsetting 
with_mutation_mask <- ifelse(maf_object@data$Hugo_Symbol == 'TTN', TRUE, FALSE)
with_mutation_barcode <- maf_object@data$Tumor_Sample_Barcode[metastatic_mask]

TTN_muts <- subsetMaf(maf = maf_object,
                            tsb = with_mutation_barcode)

#Use "in" to find patients with TNN
rna_clinical$mutation_status <- ifelse(rna_clinical$Tumor_Sample_Barcode %in% TTN_muts@data$Tumor_Sample_Barcode, TRUE, FALSE)

#Creating table
contig <- table(rna_clinical$metastasis, rna_clinical$mutation_status)

contig

#Fisher's Exact Test
fisher_test <- fisher.test(contig)
fisher_test

#Mosaic Plot
mosaicplot(contig)

```

```{r}
#5 Contingency table for nonmetastatic patients 

rna_clinical$metastasis <- factor(rna_clinical$metastasis, levels=c('Metastatic', 'Primary solid Tumor'))

nonmetastatic_mask <- ifelse(rna_clinical$metastasis == 'Primary solid Tumor', TRUE, FALSE)
nonmetastatic_barcode <- maf_object@clinical.data$Tumor_Sample_Barcode[nonmetastatic_mask]

#Subsetting 
with_mutation_mask <- ifelse(maf_object@data$Hugo_Symbol == 'TTN', TRUE, FALSE)
with_mutation_barcode <- maf_object@data$Tumor_Sample_Barcode[nonmetastatic_mask]

TTN_muts <- subsetMaf(maf = maf_object,
                            tsb = with_mutation_barcode)

#Use "in" to find patients with TTN
rna_clinical$mutation_status <- ifelse(rna_clinical$Tumor_Sample_Barcode %in% TTN_muts@data$Tumor_Sample_Barcode, TRUE, FALSE)

#Creating table
contig <- table(rna_clinical$metastasis, rna_clinical$mutation_status)

contig

#Fisher's Exact Test
fisher_test <- fisher.test(contig)
fisher_test

#Mosaic Plot
mosaicplot(contig)







```






```{r}
#6 Volcano plot 

clin_query <- GDCquery(project = "TCGA-SKCM",
data.category = "Clinical",
data.type = "Clinical Supplement",
data.format = 'BCR Biotab')
GDCdownload(clin_query)
clinical.BRCtab.all <- GDCprepare(clin_query)

rad <- clinical.BRCtab.all$clinical_radiation_skcm[-c(1,2),]
drug <- clinical.BRCtab.all$clinical_drug_skcm[-c(1,2),]


library(SummarizedExperiment)
BiocManager::install('DESeq2')
library(DESeq2)

rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)

rna_clinical$Tumor_Sample_Barcode <- rna_clinical$patient

rna_clinical$metastasis <- rna_se@colData$definition

#Mask out patients who are not "Metastatic" or "Primary solid Tumor"
#metastatic_mask <- ifelse(rna_clinical$metastasis == 'Metastatic', TRUE, #ifelse(rna_clinical$metastasis == 'Primary solid Tumor', TRUE, FALSE))

#metastasis_na_mask <- !is.na(rna_clinical$metastasis)
#rna_clinical <- rna_clinical[metastasis_na_mask,]

#rna_counts <-rna_counts [, metastasis_na_mask]




#Change rna_genes from s4 object to data frame
rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)

#Creating a new dataframe with data from unstranded 
rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)

#Making row names informative for rna_genes
rownames(rna_genes) <- rna_genes$gene_id

#Making row and column names informative for rna_counts
rownames(rna_counts) <- rna_genes$gene_id
colnames(rna_counts) <- rownames(rna_clinical)

#Subset normal tissue samples from rna_clinical and rna_counts

unique(rna_clinical$definition)

normal_mask <- ifelse(rna_clinical$metastasis == "Solid Tissue Normal", F, T)

rna_clinical <- rna_clinical[normal_mask,]

rna_counts <- rna_counts[, normal_mask]


#Categorical variable: Metastatic vs nonmetastatic
#Covariance variables: treatment effects (radiation, chemotherapy, immunotherapy, molecular therapy, vaccine), race, gender, vital status

#Hot encoding for treatment effects


#subset drug dataframe for each treatment 

radiation_mask <- ifelse(rad$bcr_patient_barcode %in% rna_clinical$Tumor_Sample_Barcode, T,F)
radiation_barcodes <- rad$bcr_patient_barcode[radiation_mask]

radiation<- as.data.frame(radiation_barcodes)

rna_clinical$radiation <- ifelse(rna_clinical$Tumor_Sample_Barcode %in% radiation$radiation_barcodes, 1, 0)


chemotherapy_mask <- ifelse(drug$pharmaceutical_therapy_type == 'Chemotherapy', TRUE, FALSE)
chemotherapy_barcodes <- drug$bcr_patient_barcode[chemotherapy_mask]

chemotherapy<- as.data.frame(chemotherapy_barcodes)

rna_clinical$chemotherapy <- ifelse(rna_clinical$Tumor_Sample_Barcode %in% chemotherapy$chemotherapy_barcodes, 1, 0)


vaccine_mask <- ifelse(drug$pharmaceutical_therapy_type == 'Vaccine', TRUE, FALSE)
vaccine_barcodes <- drug$bcr_patient_barcode[vaccine_mask]

vaccine<- as.data.frame(vaccine_barcodes)

rna_clinical$vaccine <- ifelse(rna_clinical$Tumor_Sample_Barcode %in% vaccine$vaccine_barcodes, 1, 0)


immunotherapy_mask <- ifelse(drug$pharmaceutical_therapy_type == 'Immunotherapy', TRUE, FALSE)
immunotherapy_barcodes <- drug$bcr_patient_barcode[immunotherapy_mask]

immunotherapy <- as.data.frame(immunotherapy_barcodes)

rna_clinical$immunotherapy <- ifelse(rna_clinical$Tumor_Sample_Barcode %in% immunotherapy$immunotherapy_barcodes, 1, 0)


molecular_mask <- ifelse(drug$pharmaceutical_therapy_type == 'Targeted Molecular therapy', TRUE, FALSE)
molecular_barcodes <- drug$bcr_patient_barcode[molecular_mask]

targeted_molecular_therapy <- as.data.frame(molecular_barcodes)

rna_clinical$targeted_molecular_therapy <- ifelse(rna_clinical$Tumor_Sample_Barcode %in% targeted_molecular_therapy$molecular_barcodes, 1, 0)


#Removing NAs from all variables in rna_clinical


gender_na_mask <- !is.na(rna_clinical$gender)
rna_clinical <- rna_clinical[gender_na_mask, ]

rna_counts <- rna_counts[,gender_na_mask]



vital_status_na_mask <- !is.na(rna_clinical$vital_status)
rna_clinical <- rna_clinical[vital_status_na_mask, ]

rna_counts <- rna_counts[,vital_status_na_mask]



race_na_mask <- !is.na(rna_clinical$race)
rna_clinical <- rna_clinical[race_na_mask,] 

rna_counts <- rna_counts[,race_na_mask]



#Removing genes with less than 20 patient samples

# use rowSums() to create a list with the total number of counts of each gene
row_sums <- rowSums(rna_counts, na.rm=TRUE)

# create a boolean mask where genes with < 20 total counts are FALSE, and genes with >= 10 total counts are TRUE  
low_counts_mask <- ifelse(row_sums >= 20, TRUE, FALSE)

# rewrite the rna_counts df, subsetting for only genes with >= 20 total counts
rna_counts <- rna_counts[low_counts_mask, ]

#update rna_genes with the low_counts_mask
rna_genes <- rna_genes$gene_id[low_counts_mask]



#Factor variables

rna_clinical$metastasis <- factor(rna_clinical$metastasis)
head(rna_clinical$metastasis)

rna_clinical$vital_status <- factor(rna_clinical$vital_status)
head(rna_clinical$vital_status)

rna_clinical$gender <- factor(rna_clinical$gender)
head(rna_clinical$gender)

rna_clinical$race <- factor(rna_clinical$race)
head(rna_clinical$race)

rna_clinical$radiation <- factor(rna_clinical$radiation)

rna_clinical$chemotherapy <- factor(rna_clinical$chemotherapy)

rna_clinical$immunotherapy <- factor(rna_clinical$immunotherapy)

rna_clinical$vaccine <- factor(rna_clinical$vaccine)

rna_clinical$targeted_molecular_therapy <- factor(rna_clinical$targeted_molecular_therapy)


#Running DESeq2

?DESeqDataSetFromMatrix
dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                             colData = rna_clinical,
                             design = ~immunotherapy + gender + race + vital_status + radiation + chemotherapy + vaccine + targeted_molecular_therapy + metastasis) 

?DESeq
dds_obj <- DESeq(dds) # note: this will likely take a long time (ie 45 minutes to 2 hours)

?resultsNames
resultsNames(dds_obj)  # see what comparisons got run

# get the young vs. old comparison
#?results
results <- results(dds_obj, format = "DataFrame", contrast = c("metastasis", 'Primary solid Tumor', 'Metastatic')) # this is case sensitive so be careful to match it with your age_category factors closely!


BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)

#Create Volcano Plot

 EnhancedVolcano(results,
    lab = rownames(results),
    x = 'log2FoldChange',
    y = 'pvalue',
    pCutoff = 0.05,
    FCcutoff = 1)


```


